### PR TITLE
Make email preferences load the client-side app with proper data

### PR DESCRIPTION
**Pull request by acelaya**
Part of https://github.com/hypothesis/lms/issues/5812

This PR updates the server template so that it injects client-side config for the email preferences app to load, as we do with the file picker, assignment launch, etc.

~The logic to save preferences is not implemented here, but in https://github.com/hypothesis/lms/pull/5902~ EDIT.: Both PRs are now merged together here.

### Testing steps

1. Check out this branch
2. Run `make shell`. Once in, run `request.find_service(s.EmailPreferencesService).preferences_url("9ea84c4bcfd9329b4495f5b27d8cb6@lms.hypothes.is", "")`.
  It should generate a URL.
3. Visit that URL. You should now see the client-side preact app, but the proper checkboxes should be selected.
4. Try changing some values and saving. It should reload the page and save the values.
5. Reloading the page should be possible, as the URLs handled by the client and the server match.

### CODE REVIEW

File: lms/static/scripts/frontend_apps/components/EmailNotificationsPreferences.tsx [hunk 1]
```diff
@@ -6,13 +6,13 @@ import { useCallback } from 'preact/hooks';
...
-  ['instructor_email_digests.days.thu', 'Thursday'],
-  ['instructor_email_digests.days.fri', 'Friday'],
-  ['instructor_email_digests.days.sat', 'Saturday'],
+  ['sun', 'Sunday'],
+  ['mon', 'Monday'],
+  ['tue', 'Tuesday'],
+  ['wed', 'Wednesday'],
+  ['thu', 'Thursday'],
+  ['fri', 'Friday'],
+  ['sat', 'Saturday'],
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/components/EmailNotificationsPreferences.tsx:15]: It turns out that in the end, the BE does not expose the `instructor_email_digests.days.` prefix, which is nicer.

I missed the discussion that happened in the original PR around not exposing this.

File: lms/static/scripts/frontend_apps/index.tsx [hunk 2]
```diff
@@ -39,10 +39,13 @@ export function init() {
...
-    VitalSourceService,
-    new VitalSourceService({ authToken: config.api.authToken })
-  );
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  if (config.api?.authToken) {
+    services.set(
+      VitalSourceService,
+      new VitalSourceService({ authToken: config.api.authToken })
+    );
+  }
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/index.tsx:48]: I'm not sure what's the right way to handle this.

We have defined the `api` prop to be mandatory in config, but for the email-notifications mode it is not present.

Changing the signature has a lot of implications, so I don't think we want to go that way.

File: lms/views/email.py [hunk 3]
```diff
@@ -77,9 +78,12 @@ def preferences_redirect(self):
     )
     def preferences(self):
         return {
-            "preferences": self.email_preferences_service.get_preferences(
-                self.request.authenticated_userid
-            ).days(),
+            "jsConfig": {
+                "mode": JSConfig.Mode.EMAIL_NOTIFICATIONS,
```
Comment (acelaya) [inline lms/views/email.py:82]: I'm not sure if importing `JSConfig` here is wrong. We could perhaps extract the `Mode` enum to use in `JSConfig` and here.

File: lms/static/scripts/frontend_apps/index.tsx (hunk 2 above)
Comment (robertknight) [reply lms/static/scripts/frontend_apps/index.tsx:48]: The `useConfig` hook has a way to specify that certain optional properties are required, so this means that accurately typing `api` to reflect that it is not present in all modes, is not too difficult.

```diff
diff --git a/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx b/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx
index 4f1cacfb..963ef912 100644
--- a/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx
+++ b/lms/static/scripts/frontend_apps/components/BasicLTILaunchApp.tsx
@@ -55,7 +55,7 @@ export default function BasicLTILaunchApp() {
     // Content URL to show in the iframe.
     viaUrl: viaURL,
     canvas,
-  } = useConfig(['hypothesisClient']);
+  } = useConfig(['api', 'hypothesisClient']);
 
   const clientRPC = useService(ClientRPC);
 
diff --git a/lms/static/scripts/frontend_apps/components/ContentSelector.tsx b/lms/static/scripts/frontend_apps/components/ContentSelector.tsx
index 71fab89c..55ca683f 100644
--- a/lms/static/scripts/frontend_apps/components/ContentSelector.tsx
+++ b/lms/static/scripts/frontend_apps/components/ContentSelector.tsx
@@ -91,7 +91,7 @@ export default function ContentSelector({
       },
       youtube: { enabled: youtubeEnabled },
     },
-  } = useConfig(['filePicker']);
+  } = useConfig(['api', 'filePicker']);
 
   // Map the existing content selection to a dialog type and value. We don't
   // open the corresponding dialog immediately, but do pre-fill the dialog
diff --git a/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx b/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx
index 09da619b..3f8b55c1 100644
--- a/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx
+++ b/lms/static/scripts/frontend_apps/components/FilePickerApp.tsx
@@ -105,7 +105,7 @@ export async function loadFilePickerConfig(
     throw new Error('Assignment editing config missing');
   }
 
-  const authToken = config.api.authToken;
+  const authToken = config.api!.authToken;
   const { path, data } = config.editing.getConfig;
   const { assignment, filePicker } = await apiCall<Partial<ConfigObject>>({
     authToken,
@@ -166,7 +166,7 @@ export default function FilePickerApp({ onSubmit }: FilePickerAppProps) {
     },
     assignment,
     filePicker: { deepLinkingAPI, formAction, formFields, promptForTitle },
-  } = useConfig(['filePicker']);
+  } = useConfig(['api', 'filePicker']);
 
   // Currently selected content for assignment.
   const [content, setContent] = useState<Content | null>(
diff --git a/lms/static/scripts/frontend_apps/components/GradingControls.tsx b/lms/static/scripts/frontend_apps/components/GradingControls.tsx
index 3bcf9c9c..7215bf96 100644
--- a/lms/static/scripts/frontend_apps/components/GradingControls.tsx
+++ b/lms/static/scripts/frontend_apps/components/GradingControls.tsx
@@ -39,7 +39,7 @@ export default function GradingControls({
 }: GradingControlsProps) {
   const {
     api: { authToken, sync: syncAPICallInfo },
-  } = useConfig();
+  } = useConfig(['api']);
 
   const clientRPC = useService(ClientRPC);
   const [hasUnsavedChanges, setHasUnsavedChanges] = useState(false);
diff --git a/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx b/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx
index ae48f910..e95d2931 100644
--- a/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx
+++ b/lms/static/scripts/frontend_apps/components/GroupConfigSelector.tsx
@@ -153,7 +153,7 @@ export default function GroupConfigSelector({
     product: {
       api: { listGroupSets: listGroupSetsAPI },
     },
-  } = useConfig();
+  } = useConfig(['api']);
 
   const useGroupSet = groupConfig.useGroupSet;
   const groupSet = useGroupSet ? groupConfig.groupSet : null;
diff --git a/lms/static/scripts/frontend_apps/config.ts b/lms/static/scripts/frontend_apps/config.ts
index 1452af00..656b45a4 100644
--- a/lms/static/scripts/frontend_apps/config.ts
+++ b/lms/static/scripts/frontend_apps/config.ts
@@ -236,7 +236,9 @@ export type EmailNotificationsPreferences = Record<WeekDay, boolean>;
 export type ConfigObject = {
   // Configuration present in all modes.
   mode: AppMode;
-  api: {
+
+  // Present in most modes.
+  api?: {
     authToken: string;
 
     // Only present in "basic-lti-launch" mode.
diff --git a/lms/static/scripts/frontend_apps/index.tsx b/lms/static/scripts/frontend_apps/index.tsx
index a0853c01..77dffff2 100644
--- a/lms/static/scripts/frontend_apps/index.tsx
+++ b/lms/static/scripts/frontend_apps/index.tsx
@@ -47,7 +47,7 @@ export function init() {
     );
   }
 
-  if (config.rpcServer && config.hypothesisClient) {
+  if (config.api && config.rpcServer && config.hypothesisClient) {
     const { authToken } = config.api;
     const clientRPC = new ClientRPC({
       authToken,
diff --git a/lms/static/scripts/frontend_apps/utils/api.ts b/lms/static/scripts/frontend_apps/utils/api.ts
index abf2d24d..980bc78f 100644
--- a/lms/static/scripts/frontend_apps/utils/api.ts
+++ b/lms/static/scripts/frontend_apps/utils/api.ts
@@ -173,7 +173,7 @@ export function useAPIFetch<T = unknown>(
 ): FetchResult<T> {
   const {
     api: { authToken },
-  } = useConfig();
+  } = useConfig(['api']);
 
   const fetcher: Fetcher<T> | undefined = path
     ? signal =>
```

**Review by robertknight — COMMENTED** [review]
I'll let @seanh advise on the backend part. The frontend changes are good except for handling the optional-ness of the `api` field in `ConfigObject`. I added a suggestion for how to deal with that.

File: lms/views/email.py (hunk 3 above)
Comment (seanh) [reply lms/views/email.py:82]: I wouldn't bother with `JSConfig`. I would just do this:

```python
return {
    "jsConfig": {
        "mode": "email-notifications",
        "emailNotifications": self.email_preferences_service.get_preferences(
            self.request.authenticated_userid
        ).days(),
    }
}
```

I'd remove `JSConfig.Mode.EMAIL_NOTIFICATIONS` from the `JSConfig` class. There's just no need for these email preferences views to be entangled with that class.

**Comment by acelaya** [comment]
> 3\. **Expected result:** as on `main` I should be redirected to `/email/preferences` (no `token`), which is a cookie-authenticated page showing the email preferences form. When I click **Save** my preferences should be saved to the DB (this is currently working on `main`).
>     **Actual result:** I get redirected to `/app/email-notifications` which is not covered by the cookie so authentication is lost. If I reload the page I will get a 404, which is both because the authentication cookie has been lost and because the backend is unaware of the `/app/email-notification` path. Clicking **Save** doesn't do anything.

Yes, this is explained in the PR description.

* The logic to save is implemented here: https://github.com/hypothesis/lms/pull/5902
* The inconveniences caused by the client-side routing to `/app/email-notifications` are also mentioned there.
  This is how we handle client apps, but it's something I'm already looking into.

File: lms/views/email.py (hunk 3 above)
Comment (acelaya) [reply lms/views/email.py:82]: Perfect. I'll change it.

File: lms/static/scripts/frontend_apps/test/index-test.js [hunk 4]
```diff
@@ -50,16 +50,20 @@ describe('LMS frontend entry', () => {
     $imports.$restore();
   });
 
-  it('renders root component', () => {
-    init();
+  ['basic-lti-launch', 'email-notifications'].forEach(mode => {
```
Comment (acelaya) [inline lms/static/scripts/frontend_apps/test/index-test.js:53]: This test remains unchanged, but I ensured it tests for the two "types" of app modes, those which route is just the mode with `/app/` prefix, and those that have a completely different URL.

**Comment by acelaya** [comment]
I have updated this PR so that the `email-notifications` client app mode resolves the same URL used by the server.

This is more convenient, as it allows for the page to be reloaded, among other things.

**Review by robertknight — COMMENTED** [review]
I think we should in future change the frontend to avoid updating the URL after the initial render, however some exploration is needed to find the best way to do that, so the workaround implemented here is appropriate for this PR.

**Comment by acelaya** [comment]
I closed this PR by mistake. When I was going to run `git push ... -f` I run `git push ... -d` instead by mistake, which deleted the remote branch.

I pushed the branch again, but GitHub did not allow this PR to be reopened, so I created https://github.com/hypothesis/lms/pull/5905 instead.

**Pull request closed without merging**